contracts/cw20-merkle-airdrop/src/contract.rs

```diff
@@ -0,0 +1,2160 @@
+
+    // Update total claimed to reflect
+    let mut claimed_amount = STAGE_AMOUNT_CLAIMED.load(deps.storage, stage)?;
+    claimed_amount += amount;
+    STAGE_AMOUNT_CLAIMED.save(deps.storage, stage, &claimed_amount)?;
+
+    // we stop all airdrops at the date of expiration, and we use the very same date
+    // as a start timestamp for vesting. if expiration is not set, vesting will not work.
+    let vesting_start_time =
+        match expiration {
+            Expiration::AtHeight(_) => {
+                return Err(ContractError::Vesting {
+                    description: "Vesting must be scheduled from timestamp, not block height"
+                        .to_string(),
+                })
+            }
+            Expiration::Never {} => return Err(ContractError::Vesting {
+                description:
+                    "Vesting must be scheduled from some timestamp, but there wasn't any provided"
+                        .to_string(),
+            }),
+            Expiration::AtTime(timestamp) => timestamp,
+        };
+
+    let transfer_message = Cw20Contract(credits_address.clone())
+        .call(Cw20ExecuteMsg::Transfer {
+            recipient: info.sender.to_string(),
+            amount,
+        })
+        .map_err(ContractError::Std)?;
+    let vesting_message = WasmMsg::Execute {
+        contract_addr: credits_address.to_string(),
+        msg: to_binary(&AddVesting {
+            address: info.sender.to_string(),
+            amount,
+            start_time: vesting_start_time.nanos(),
+            duration: VESTING_DURATION,
+        })?,
+        funds: vec![],
+    };
+    let res = Response::new()
+        .add_message(transfer_message)
+        .add_message(vesting_message)
+        .add_attributes(vec![
+            attr("action", "claim"),
+            attr("stage", stage.to_string()),
+            attr("address", info.sender.to_string()),
+            attr("amount", amount),
+        ]);
+    Ok(res)
+}
+
+pub fn execute_withdraw_all(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+) -> Result<Response, ContractError> {
+    // authorize owner
+    let cfg = CONFIG.load(deps.storage)?;
+    let owner = cfg.owner.ok_or(ContractError::Unauthorized {})?;
+    if info.sender != owner {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    if !STAGE_PAUSED.load(deps.storage, 1)?
+        && env.block.time
+            <= match STAGE_EXPIRATION.load(deps.storage, 1)? {
+                Expiration::AtTime(timestamp) => timestamp,
+                _ => {
+                    return Err(ContractError::Std(StdError::generic_err(
+                        "withdraw_all only works if AtTime expiration is set",
+                    )))
+                }
+            }
+            .plus_nanos(VESTING_DURATION)
+    {
+        return Err(ContractError::Std(StdError::generic_err(
+            "withdraw_all only works 3 months after the end of the event",
```

Comment:
> why do we allow custom expiration in the `execute_register_merkle_root` whereas we strictly claim here that it's 3 months?

@oopcode asked me to introduce as little patches as possible, hence you can see these shenanigans…
Actually, there is the same problem with airdrop stages, contract has support for them, but only works for stage 1.

Reply:
I don't really like when there is a code that's never meant to be used because it complicates code review very much for no reason actually. but okay, if you've been directly asked for it, that's kind of fine

Reply:
okay, I see. but still the hardcoded value in the error message is not good. imagine if we decide to use 90 days instead of 3 months. we'll have to take care not only about the VESTING_DURATION const, but also all errors related to it. this is one of the reasons why an error without any mention of concrete duration is good imo

Reply:
imo can be done in a separate PR. marking it with this msg to create a list of such complaints a bit later

Reply:
Done!